### PR TITLE
Socket initialization

### PR DESCRIPTION
**Pull request by CIPop**
Socket initialization in System.Net.Sockets will depend on System.Net.NameResolution (Windows only).
Extracted and refactored IP protocol version detection.

### CODE REVIEW

**Comment by CIPop** [comment]
@pgavlin PTAL

File: src/System.Net.NameResolution/src/System/Net/DNS.cs [hunk 1]
```diff
@@ -157,6 +157,7 @@ internal static IPHostEntry InternalGetHostByAddress(IPAddress address, bool inc
         public static string GetHostName()
         {
             GlobalLog.Print("Dns.GetHostName");
+            NameResolutionPal.EnsureSocketsAreInitialized();
```
Comment (pgavlin) [inline src/System.Net.NameResolution/src/System/Net/DNS.cs]: Out of curiosity, why move these calls out of the PAL? Personally, I'd leave it up to the PAL implementation to check for any required initialization and leave it out of the up-stack code.

Comment (CIPop) [inline src/System.Net.NameResolution/src/System/Net/DNS.cs]: The main reason for moving was the calls to Socket.OSSupportsIPv6 which used to force the initialization but now it expects it has been done already. 
I agree that it make sense for these to be in PAL: that way the *NIX PALs don't even need to make the extra (no-op) call. (Fixing.)

Comment (CIPop) [inline src/System.Net.NameResolution/src/System/Net/DNS.cs]: @pgavlin Verified and this is not entirely possible as the PAL code was written.
For example, in any of the Async/APM APIs, the calls need to be made before Socket.OSSupportsIPv6.

**Pull request merged**